contracts/exchange/test/transactions_unit_tests.ts

```diff
@@ -0,0 +1,694 @@
+
+            const callData = getExecutableCallData(
+                true,
+                getExecuteTransactionCallData(transaction1, validSignature),
+                '0xdeadbeef',
+            );
+
+            const transaction2 = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                data: callData,
+                signerAddress: accounts[1],
+                domain,
+            };
+            const transactionHash2 = transactionHashUtils.getTransactionHashHex(transaction2);
+            const abiEncoder = AbiEncoder.createMethod('TransactionError', ['uint8', 'bytes32']);
+            const errorData = abiEncoder.encode([
+                ExchangeRevertErrors.TransactionErrorCode.NoReentrancy,
+                transactionHash1,
+            ]);
+
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(transactionHash2, errorData);
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction2, validSignature, {
+                from: accounts[1], // Different then the signing addresses
+            });
+            return expect(tx).to.revertWith(expectedError);
+        });
+
+        it('should revert if the transaction has been executed previously', async () => {
+            const validSignature = randomSignature();
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const transaction = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                data: getExecutableCallData(true, constants.NULL_BYTES, constants.NULL_BYTES),
+                domain,
+            };
+            const transactionHash = transactionHashUtils.getTransactionHashHex(transaction);
+
+            // Use the transaction in execute transaction.
+            await expect(
+                transactionsContract.executeTransaction.sendTransactionAsync(transaction, validSignature),
+            ).to.be.fulfilled('');
+
+            // Use the same transaction to make another call
+            const expectedError = new ExchangeRevertErrors.TransactionError(
+                ExchangeRevertErrors.TransactionErrorCode.AlreadyExecuted,
+                transactionHash,
+            );
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction, validSignature);
+            return expect(tx).to.revertWith(expectedError);
+        });
+
+        it('should revert if the signer != msg.sender and the signature is not valid', async () => {
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const transaction = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
```

Comment:
This is getting a little repetitive. How about having `EMPTY_ZERO_EX_TRANSACTION`  hold some timestamp in the distant future? Like `new BigNumber(Math.floor(_.now() / 1000) + 86400)`.